### PR TITLE
Add focus trapping to Modal component

### PR DESCRIPTION
**Pull request by petesfrench**
## Done

- Limits movement with the tab key to within the modal.

### Storybook

To see rendered examples of all react-components, run:

```shell
yarn start
```

### QA in your project

from `react-components` run:

```shell
yarn build
npm pack
```

Install the resulting tarball in your project with:

```shell
yarn add <path-to-tarball>
```

### QA steps

- Check that you can't tab outside of the modal and that the it moves appropriately.

## Fixes

Fixes: https://github.com/canonical-web-and-design/react-components/issues/591

### CODE REVIEW

File: src/components/Modal/Modal.tsx [hunk 1]
```diff
@@ -45,12 +45,8 @@ export const Modal = ({
   // anywhere else on the page or hits the escape key
   useEffect(() => {
     const keyDown = (e) => {
-      if (e.code === "Escape") {
-        // Close panel if Esc keydown detected
-        if (close) {
-          close();
-        }
-      }
+      const listener = keyListenersMap.get(e.keyCode);
```
Comment (hatched) [inline src/components/Modal/Modal.tsx:48]: `keyListenersMap` is being used before it's defined.

File: src/components/Modal/Modal.tsx [hunk 2]
```diff
@@ -59,13 +55,40 @@ export const Modal = ({
     return () => {
       document.removeEventListener("keydown", keyDown);
     };
-  }, [close]);
+  });
+
+  const modalRef = React.useRef();
+  const handleTabKey = (e) => {
+    const focusableModalElements = modalRef.current.querySelectorAll(
```
Comment (hatched) [inline src/components/Modal/Modal.tsx:62]: It's possible that this returns undefined so you'll need to handle that eventuality. I'm surprised the TS compiler didn't complain about this. You might want to check that you have error reporting enabled in your editor.

File: src/components/Modal/Modal.tsx [hunk 3]
```diff
@@ -59,13 +55,40 @@ export const Modal = ({
...
+      'a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select'
```
Comment (hatched) [inline src/components/Modal/Modal.tsx:63]: IIRC you can also focus any element with a `tabindex`. You can also focus anything with `contentEditable`.
This also doesn't include disabled elements.

I found this SO answer that might provide a more robust selector: https://stackoverflow.com/a/30753870/3732840

File: src/components/Modal/Modal.tsx [hunk 4]
```diff
@@ -59,13 +55,40 @@ export const Modal = ({
...
+    [27, close],
+    [9, handleTabKey],
+  ]);
 
   return (
     <div
       className={classNames("p-modal", className)}
       onClick={close}
       {...wrapperProps}
+      role="dialog"
```
Comment (hatched) [inline src/components/Modal/Modal.tsx:90]: This will have duplicate nested elements with the role `dialog` (see the `section` element below).

**Review by hatched — COMMENTED** [review]
Understanding that this is still a WIP I've added a few comments.

You'll also want to add extensive tests as this type of feature is a prime andidate for edge case issues.

**Comment by petesfrench** [comment]
@hatched Thank you for the feedback and a very good point on testing.

File: src/components/Modal/Modal.test.tsx [hunk 5]
```diff
@@ -57,4 +51,51 @@ describe("Modal ", () => {
...
+
+    expect(closeButton).toBe(document.activeElement);
+
+    userEvent.tab();
+
+    expect(cancelButton).toBe(document.activeElement);
+
+    userEvent.tab();
+
+    console.log(closeButton);
```
Comment (bartaz) [inline src/components/Modal/Modal.test.tsx:98]: This is unnecessary
```suggestion
```

File: src/components/Modal/Modal.tsx [hunk 6]
```diff
@@ -38,34 +44,67 @@ export const Modal = ({
   title,
   ...wrapperProps
 }: Props): ReactElement => {
+  const focusableElementSelectors =
+    'a[href]:not([tabindex="-1"]), button:not([disabled]), textarea:not([disabled]):not([tabindex="-1"]), input:not([disabled]):not([tabindex="-1"]), select:not([disabled]):not([tabindex="-1"]), area[href]:not([tabindex="-1"]), iframe:not([tabindex="-1"]), [tabindex]:not([tabindex="-1"]), [contentEditable=true]:not([tabindex="-1"])';
```
Comment (bartaz) [inline src/components/Modal/Modal.tsx:48]: Is this based on any documentation or stackoverflow post? would be good to comment with a link to that.

File: src/components/Modal/Modal.tsx [hunk 7]
```diff
@@ -38,34 +44,67 @@ export const Modal = ({
...
   const descriptionId = useRef(nanoid());
   const titleId = useRef(nanoid());
 
-  // This useEffect sets up listeners so the panel will close if user clicks
-  // anywhere else on the page or hits the escape key
+  const modalRef: MutableRefObject<HTMLElement> = useRef(null);
+  const focusableModalElements = useRef(null);
+  const handleTabKey = (e: KeyboardEvent<HTMLDivElement>) => {
```
Comment (bartaz) [inline src/components/Modal/Modal.tsx:54]: Out of curiosity - is this neccessary to completely override Tabs mechanism in modal?

As far as I remember in example in Vanilla we simply catch tabbing out from last focusable element and redirect focus to first one. Here it seems we are reimplementing tab functionality completely.

File: src/components/Modal/Modal.test.tsx [hunk 8]
```diff
@@ -24,14 +16,16 @@ describe("Modal ", () => {
...
   });
 
   it("displays any content in the button row", () => {
     const wrapper = shallow(
       <Modal buttonRow={<div>I am not a button</div>}>Button tester</Modal>
     );
-    expect(wrapper).toMatchSnapshot();
+    expect(wrapper.text()).toContain("I am not a button");
+    expect(wrapper.text()).toContain("Button tester");
+    expect(wrapper.text()).not.toContain("I am something else");
```
Comment (hatched) [inline src/components/Modal/Modal.test.tsx:28]: I'm not entirely sure but I think you've added a 'falsy' test case to ensure that the `toContain` assertions aren't just arbitrarily passing. Laudable goal but in this case it's not really necessary as you're effectively testing that the assertion library is working....and well....if it's not we have bigger problems 😆 

Where you'd typically want to do this is if the Modal would have the string "I am something else" in the event that nothing was defined with `buttonRow`. Then you'd want to make sure it didn't exist vs just being appended to.

File: src/components/Modal/Modal.tsx (hunk 6 above)
Comment (bartaz) [reply src/components/Modal/Modal.tsx:48]: @sowasred2012 You resolved this one, but I don't see any mention where does this selector come from?

**Review by bartaz — APPROVED** [review]
Looks good. Just a small comment about link to docs that I mentioned before.

**Comment by sowasred2012** [comment]
@bartaz yep, sorry - I'd fixed it locally, then spoke to you about the other issue before actually pushing it - it's there now.
